### ISSUE TITLE
[npm 3] Get the packager and CLI to work with npm 3

### ISSUE THREAD

**Issue by ide** [facebook/react-native#3317]
After npm installing on master (Node 4.1.2, npm 3.3.6) running the packager works fine but trying to load the UIExplorer prints lots of messages like "Unable to resolve module LinkedStateMixin". These unresolved modules do exist under node_modules, so my suspicion is that these files are blacklisted and npm 3's deduping removed the non-blacklisted copy. If this is actually the case and npm 3 is deterministic, then we could blacklist the files that are getting deduped away.

**Comment by ajwhite** [comment]
Didn't have much to do tonight so I took a look at this, it looks like it's more of a problem with Watchman, and perhaps how npm has structured things (although I can't see any visual differences there).

On `npm 3.3.6`, [I see this output from repeated 5 times](https://github.com/facebook/watchman/blob/11b1f4abebebf4893e294c85314e27f80d9aa71d/watcher/fsevents.c#L273-L279):
```
[1:03:20 AM] <END>   Loading bundles layout (0ms)

React packager ready.

Recrawled this watch 1 times, most recently because:
/Users/ajwhite/Development/Labs/react-native: kFSEventStreamEventFlagUserDropped
To resolve, please review the information on
https://facebook.github.io/watchman/docs/troubleshooting.html#recrawl
To clear this warning, run:
`watchman watch-del /Users/ajwhite/Development/Labs/react-native ; watchman watch-project /Users/ajwhite/Development/Labs/react-native`
```

I pulled up a couple `npm` versions to compare the difference between running the packager on `npm 3.3.6` and `npm 2.14.4` (my default). It became clear that **watchman was not fully scanning node_modules directory trees**. 

I was able to narrow it down to looking at the output in [crawlers/watchman.js](https://github.com/facebook/react-native/blob/master/packager/react-packager/src/DependencyResolver/crawlers/watchman.js#L56) where the files get picked up. I [updated that line to](https://github.com/ajwhite/react-native/blob/4a9ff583003bef0b4a9d6e921ea0c78a23ce6bb8/packager/react-packager/src/DependencyResolver/crawlers/watchman.js#L56-L58):
```js
//..
if (filePath.indexOf('react-tools') > -1) {
  console.log('react-tools file:', filePath);
}
//..
```

This yields two very different results on `npm 3.3.6` and `2.14.4`. Here's the comparison of the files that watchman was able to discover between the two in `node_modules/react-tools`:


### Failing version, npm 3.3.6
You'll notice it never goes past
`/Users/ajwhite/Development/Labs/react-native/node_modules/react-tools/node_modules/commoner/node_modules/` and ends up skipping all the import files required for the [ResultionRequest.resolveDependency](https://github.com/facebook/react-native/blob/master/packager/react-packager/src/DependencyResolver/DependencyGraph/ResolutionRequest.js#L84-L85) method to discover.

Put in gist to avoid 100 lines:
https://gist.github.com/ajwhite/c82ac755389e62350657


### Working version, npm 2.14.4
Here, you'll see that they're all discovered

https://gist.github.com/ajwhite/c82ac755389e62350657

----

This can be reproduced on my [npm3-investigating](https://github.com/ajwhite/react-native/tree/npm3-investigating) branch.

**TL;DR** - Nothing in `node_modules/react-tools/src` ends up getting scanned and thus never becomes added to `HasteMap`. I only looked as far as `node_modules/react-tools`, but I'd assume it is an issue for other dependencies. I'm lead to believe this may be an issue with [watchman](https://github.com/facebook/watchman/blob/11b1f4abebebf4893e294c85314e27f80d9aa71d/watcher/fsevents.c#L273-L279).

**Comment by ide** [comment]
The packager blacklists certain paths: https://github.com/facebook/react-native/blob/0ad05043f5f268c6241d34997ca8773b1e88357b/packager/blacklist.js#L15. What happens if you comment some of these out?

**Comment by ajwhite** [comment]
I fully commented those out, added others in; nothing.

Gimme 2 mins, I think I got the issue..

**Comment by ajwhite** [comment]
Ha.. gotta love this..

`watchman shutdown-server`. Then run the packager again.

@wez leads to this over in https://github.com/ember-cli/ember-cli/issues/4101#issuecomment-103299380


![solved](http://g.recordit.co/MHHzoTMqTu.gif)


:sleeping: one of those nights where you find the simple solution *after* doing everything but that.

Perhaps if that error is caught, and `npm` is `>= 3.x`, it can be warned in the console to run `watchman shutdown-server`.


![screen shot 2015-10-10 at 1 45 01 am](https://cloud.githubusercontent.com/assets/656630/10409487/8bcc1b1e-6ef0-11e5-9699-063e26813374.png)

**Comment by wez** [comment]
That recording goes by too quickly for me to see what's going on, and your commentary isn't clear to me (I don't use react-native either, so I'm at a disadvantage), so forgive me if this sounds like a stupid question:

Are you saying that the problem was resolved by restarting the watchman service?

Does the problem recur if you downgrade and upgrade npm again?
If so, could you try a HEAD install of watchman and see if the issue is resolved?
I'm thinking that maybe you hit the problem solved by https://github.com/facebook/watchman/commit/11b1f4abebebf4893e294c85314e27f80d9aa71d

**Comment by ajwhite** [comment]
I'm mobile at the moment, but I can expand upon what I had briefly described. 

The problem seemed to arise after running watchman on the project that had been installed with npm <3, and then later reinstalled with npm >=3. 

It appears to only be after the scenario of upgrading npm, and reinstalling the project with the newer version. 

This issue did not surface again after restarting watchman. That appeared to resolve the problem.

**Comment by ide** [comment]
@ajwhite Thanks for investigating. Looks like we're good with npm 3 and restarting Watchman will do the trick until the new version is published. 

Also I just tested installing the CLI tools and setting up a new project with npm 3 and the sample app worked fine.

**Comment by wez** [comment]
@ajwhite please open an issue in https://github.com/facebook/watchman/issues if watchman HEAD doesn't resolve this for you; thanks!

**Comment by blairvanderhoof** [comment]
What is the solution to this issue with LinkedStateMixin not being resolved?

**Comment by grabbou** [comment]
Well, unfortunately it does not work well with `npm3` flat structure, because it takes some `react-tools` from the packages I depend on and put in in the top level, whereas it still keeps `react-tools` inside react-native because of different version. That causes `collision name` error and makes it impossible to run packager. Any ideas?
